src/serialization/sb3.js

```diff
@@ -792,28 +855,57 @@ const deserializeFields = function (fields) {
 };
 
 /**
- * Covnert serialized INPUT and FIELD primitives back to hydrated block templates.
+ * Convert dynamic blockInfo JSON from a custom deserialization function into an XML element ready for runtime use.
+ * @param {object} dynamicBlockInfo - a BlockInfo-like object with properties like `opcode` and `isDynamic`.
+ * @returns {Element} - an XML Element with tag name 'mutation' and an attribute called 'blockInfo'
+ */
+const makeMutation = function (dynamicBlockInfo) {
+    const xmlString = `<mutation blockInfo="${xmlEscape(JSON.stringify(dynamicBlockInfo))}"/>`;
+    return mutationAdapter(xmlString);
+};
+
+/**
+ * Convert serialized INPUT and FIELD primitives back to hydrated block templates.
  * Should be able to deserialize a format that has already been deserialized.  The only
  * "east" path to adding new targets/code requires going through deserialize, so it should
  * work with pre-parsed deserialized blocks.
  *
+ * @param {!Runtime} runtime Runtime object to load all structures into.
  * @param {object} blocks Serialized SB3 "blocks" property of a target. Will be mutated.
  * @return {object} input is modified and returned
  */
-const deserializeBlocks = function (blocks) {
+const deserializeBlocks = function (runtime, blocks) {
     for (const blockId in blocks) {
-        if (!Object.prototype.hasOwnProperty.call(blocks, blockId)) {
+        if (blocks.hasOwnProperty && !blocks.hasOwnProperty(blockId)) {
             continue;
         }
-        const block = blocks[blockId];
-        if (Array.isArray(block)) {
-            // this is one of the primitives
-            // delete the old entry in object.blocks and replace it w/the
-            // deserialized object
+        let block = blocks[blockId];
+        const {extensionId, extendedOpcode} = getExtensionAndOpcode(block);
+        const extensionInfo = runtime.getExtensionInfo(extensionId);
+        const customDeserialize = getSerializationInfo(extensionInfo, extendedOpcode).deserialize;
+        if (customDeserialize) {
+            block = customDeserialize(block);
+            blocks[blockId] = block; // customDeserialize might have made a new object; make sure to keep it!
+            // the deserialize function should return opcode 'foo' and we'll fix it to 'ext_foo'
```

Comment:
Related to the comment earlier in this PR, if we're doing fix-up here anyways of what the customDeserialize function returns, I think we might as well construct the block JSON here instead of the customDeserialize function and have that function just return the blockInfo.